### PR TITLE
Fix static building for ocaml-ng

### PR DESCRIPTION
**Pull request by balsoft**
###### Motivation for this change

Currently attempting to build `pkgsStatic.ocaml` and friends fails miserably. This PR:
* Adds a static adapter that fixes `ocaml` and some subset of `ocamlPackages`
* Fixes some individual packages from `ocamlPackages`

I am not sure that the adapter *should* go in top-level/static.nix, if it belongs in development/compilers/ocaml/generic.nix, then it's pretty easy to move it there.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @vbgl @regnat

### CODE REVIEW

**Comment by vbgl** [comment]
Naive question after a quick look at the changes: why is `findlib` often both in `buildInputs` and `nativeBuildInputs`?

**Comment by balsoft** [comment]
@vbgl Because it doesn't appear to be working otherwise :)

My guess is that it needs to find both host and target libraries, thus requiring both host and target findlib.

File: pkgs/development/tools/ocaml/js_of_ocaml/3.0.nix [hunk 1]
```diff
@@ -7,7 +7,8 @@ stdenv.mkDerivation {
 
 	inherit (js_of_ocaml-compiler) version src installPhase meta;
 
-	buildInputs = [ ocaml findlib dune ocaml-migrate-parsetree ppx_tools_versioned ];
+	buildInputs = [ findlib ocaml-migrate-parsetree ppx_tools_versioned ];
+  nativeBuildInputs = [ ocaml findlib dune ];
```
Comment (matthewbauer) [inline pkgs/development/tools/ocaml/js_of_ocaml/3.0.nix]:
```suggestion
      nativeBuildInputs = [ ocaml findlib dune ];
```

Comment (balsoft) [reply pkgs/development/tools/ocaml/js_of_ocaml/3.0.nix]: Fixed.

File: pkgs/development/ocaml-modules/nocrypto/default.nix [hunk 2]
```diff
@@ -52,7 +52,7 @@ stdenv.mkDerivation rec {
   ];
 
   nativeBuildInputs = [ ocaml findlib ocamlbuild cc-wrapper ];
-  buildInputs = [ ocamlbuild findlib topkg cpuid ocb-stubblr ];
+  buildInputs = [ ocamlbuild findlib topkg cpuid ocb-stubblr ppx_deriving ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/nocrypto/default.nix]: `ppx_deriving` should be propagated.

Comment (vbgl) [reply pkgs/development/ocaml-modules/nocrypto/default.nix]: This broke `jackline`, see: https://hydra.nixos.org/build/107730405

Will fix there: https://github.com/NixOS/nixpkgs/pull/74437

**Comment by anmonteiro** [comment]
@balsoft commit 7263951cc87a9cd0bd377cef83c9241d0dffb665 mentions it fixes the static build of OCaml for versions <= 4.07. I got OCaml 4.08 and 4.09 going under musl with the following (I think they switched the build to autoconf since 4.08):

```nix
(_: super:
    {
      ocaml = super.ocaml.overrideDerivation (o: {
        preConfigure = ''
          configureFlagsArray+=("CC=$CC" "AS=$AS" "PARTIALLD=$LD -r")
        '';
        configureFlags = (lib.remove "--no-shared-libs" o.configureFlags) ++ [
          "--disable-shared"
           "--host ${o.stdenv.hostPlatform.config}"
          "--target ${o.stdenv.targetPlatform.config}"
        ];
      });
    })
```

Ideally this would make it upstream to nixpkgs -- I'm looking for pointers on how to add these flags conditionally based on the OCaml version.

**Comment by vbgl** [comment]
The `configureFlags` attribute of the derivation for the OCaml compilers has a “flag” function that should help to deal with the discrepancies between old and new configure flags:

https://github.com/NixOS/nixpkgs/blob/5f5df51bf0b0278b2226158731b7bac214962848/pkgs/development/compilers/ocaml/generic.nix#L47

**Comment by anmonteiro** [comment]
I'm happy to submit a PR adding support for OCaml >= 4.08, but I'm still having an issue where I can't use packages that build with `topkg`. I keep getting e.g.:

```

building
No such package: topkg
File "./pkg/pkg.ml", line 4, characters 5-10:
4 | open Topkg
         ^^^^^
Error: Unbound module Topkg
builder for '/nix/store/7gfy9f8ai9bwbv1q112v3d40nwgas8w4-ocaml4.09.0-fmt-0.8.5-x86_64-unknown-linux-musl.drv' failed with exit code 2
```

Any pointers on what to look for here would be appreciated. I tried messing with `nativeBuildInputs` (added both `topkg` and `findlib`) but it doesn't seem to change anything.

**Pull request merged**